WIP: Sc/mhd treemesh 3d

SimonCan opened the pull request:
This branch is a fork of the ns-treemesh-3d branch. It implements the compressible visco-resistive MHD equations in 3 dimensions on the treemesh.

This is a todo list before we can merge with the main branch:
- [x]  Perform tests on different systems, like Alfven waves, Kelvin-Helmholtz suppression and dynamos.
- [x] Write tests.
- [x] Add at least one more example elixir with source term that captures Ohmic heating. Perhaps a dynamo is a good example for this.

Convergence test for `examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl`:
```
####################################################################################################
l2
rho                 rho_v1              rho_v2              rho_v3              rho_e_total         B1                  B2                  B3                  psi                 
error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       
2.82e-05  -         2.54e-05  -         1.23e-05  -         2.86e-05  -         1.52e-05  -         2.61e-05  -         1.23e-05  -         2.86e-05  -         3.45e-06  -         
1.82e-06  3.95      1.77e-06  3.84      8.30e-07  3.88      2.00e-06  3.84      1.77e-06  3.10      1.89e-06  3.79      7.68e-07  4.00      1.98e-06  3.85      4.08e-07  3.08      
1.20e-07  3.92      2.81e-07  2.66      1.37e-07  2.60      3.15e-07  2.66      1.92e-07  3.20      1.53e-07  3.62      4.32e-08  4.15      1.52e-07  3.71      4.87e-08  3.07      

mean      3.94      mean      3.25      mean      3.24      mean      3.25      mean      3.15      mean      3.71      mean      4.08      mean      3.78      mean      3.07      
----------------------------------------------------------------------------------------------------
linf
rho                 rho_v1              rho_v2              rho_v3              rho_e_total         B1                  B2                  B3                  psi                 
error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       error     EOC       
6.44e-05  -         5.54e-05  -         2.70e-05  -         6.21e-05  -         6.13e-05  -         5.50e-05  -         2.91e-05  -         6.22e-05  -         1.20e-05  -         
4.18e-06  3.94      3.76e-06  3.88      1.98e-06  3.77      4.39e-06  3.82      6.16e-06  3.31      4.05e-06  3.76      1.88e-06  3.95      4.00e-06  3.96      1.16e-06  3.37      
2.56e-07  4.03      6.51e-07  2.53      3.85e-07  2.36      6.78e-07  2.70      6.24e-07  3.30      3.94e-07  3.36      1.21e-07  3.96      3.91e-07  3.36      1.10e-07  3.39      

mean      3.99      mean      3.21      mean      3.07      mean      3.26      mean      3.31      mean      3.56      mean      3.95      mean      3.66      mean      3.38      
----------------------------------------------------------------------------------------------------
(Dict(:l2 => [3.952415644889589 3.844680905098259 … 3.850140750933816 3.078559686325443; 3.9227371835038847 2.656199757465629 … 3.7083121872841254 3.068010438713628], :linf => [3.9449368370829605 3.883093769494825 … 3.960055037764503 3.3737554866850537; 4.032403421597966 2.529517445602747 … 3.356307225358875 3.3946538077719426]), Dict{Symbol, LinearAlgebra.Transpose{Float64, Matrix{Float64}}}(:l2 => [2.8186561413297704e-5 2.5427812774320658e-5 … 2.8594041808694996e-5 3.4469864312446496e-6; 1.8207337115709807e-6 1.769883697062282e-6 … 1.9827487935071304e-6 4.080381452736118e-7; 1.2005627824705508e-7 2.807688527216588e-7 … 1.5168914122661963e-7 4.8656133653046794e-8], :linf => [6.439065779950681e-5 5.5417382982404784e-5 … 6.224485633657059e-5 1.2003505640157272e-5; 4.18098435839287e-6 3.755937114800073e-6 … 3.999522258095467e-6 1.1579945124350865e-6; 2.5550779747529617e-7 6.5051555626916e-7 … 3.9053410692477253e-7 1.101067338920899e-7]))
  ```

andrewwinters5000 committed:
initial commit of 3d approximation for CNS

andrewwinters5000 committed:
typo fix in CNS 2D file

andrewwinters5000 committed:
add TGV elixir for compressible Navier-Stokes

jlchan committed:
add DGMulti CNS 3D elixir

andrewwinters5000 committed:
need TreeMesh versions of slip wall BC in 3D compressible Euler for convergence testing

andrewwinters5000 committed:
update comments in CNS 3D file

andrewwinters5000 committed:
add convergence test elixir for 3D compressible Navier-Stokes

andrewwinters5000 committed:
remove debug statements from new elixir

andrewwinters5000 committed:
adjust maximum tree size in new elixir

andrewwinters5000 committed:
Merge branch 'ns-treemesh-3d' of github.com:trixi-framework/Trixi.jl into ns-treemesh-3d

andrewwinters5000 committed:
add convergence test for DGMulti 3D

andrewwinters5000 committed:
add battery of 3D parabolic tests. All pass locally

andrewwinters5000 committed:
remove old comments and TODOs

andrewwinters5000 committed:
bug fix in divergence B analysis routine

ranocha committed:
Merge branch 'main' into ns-treemesh-3d

andrewwinters5000 committed:
Merge branch 'ns-treemesh-3d' of github.com:trixi-framework/Trixi.jl into ns-treemesh-3d

ranocha committed:
Merge branch 'main' into ns-treemesh-3d

andrewwinters5000 committed:
add vorticity and enstrophy functions

andrewwinters5000 committed:
add first version of an enstrophy callback

andrewwinters5000 committed:
add enstrophy callback to TGV elixir

andrewwinters5000 committed:
Merge branch 'ns-treemesh-3d' of github.com:trixi-framework/Trixi.jl into ns-treemesh-3d

andrewwinters5000 committed:
update TGV test values since switched surface flux to flux_hllc

andrewwinters5000 committed:
Merge branch 'main' into ns-treemesh-3d

jlchan committed:
adding draft DGMulti analysis callback

jlchan committed:
adding draft DGMulti analysis callback


forgot the return statement

jlchan committed:
Merge remote-tracking branch 'refs/remotes/origin/ns-treemesh-3d'

Conflicts:
	src/callbacks_step/analysis_dgmulti.jl

ranocha committed:
Merge branch 'main' into ns-treemesh-3d

andrewwinters5000 committed:
Apply suggestions from code review

Co-authored-by: Hendrik Ranocha <ranocha@users.noreply.github.com>

andrewwinters5000 committed:
make sure Experimental code text renders properly in docs

jlchan committed:
adding extra analysis integrals

jlchan committed:
Merge branch 'ns-treemesh-3d' of https://github.com/trixi-framework/Trixi.jl into ns-treemesh-3d

andrewwinters5000 committed:
remove unused file path from parabolic test scripts

jlchan committed:
remove threading for reduction

jlchan committed:
Merge branch 'main' into ns-treemesh-3d

jlchan committed:
Merge remote-tracking branch 'origin/ns-treemesh-3d' into ns-treemesh-3d

jlchan committed:
Merge branch 'main' into ns-treemesh-3d

andrewwinters5000 committed:
Merge branch 'main' into ns-treemesh-3d

andrewwinters5000 committed:
Apply suggestions from code review

Co-authored-by: Michael Schlottke-Lakemper <michael@sloede.com>

andrewwinters5000 committed:
add designation to the 2D test set

andrewwinters5000 committed:
Merge branch 'ns-treemesh-3d' of github.com:trixi-framework/Trixi.jl into ns-treemesh-3d

andrewwinters5000 committed:
Merge branch 'main' into ns-treemesh-3d

SimonCan committed:
Added CompressibleMhdDiffusion3D.

SimonCan committed:
Added compressible, viscous and diffusive MHD equations module.

SimonCan committed:
Added visco-resistive Alfven wave exmple elixir.

SimonCan committed:
Added covnergence test for the visco-resistive MHD equtions,
in te shape of linear Alfven waves.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Typos in documentation for compressible_mhd_3d.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

iomsn committed:
Added resistive and viscous MHD wave example in 3d.

iomsn committed:
Added resistive MHD test.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

iomsn committed:
Minor typo.

iomsn committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Removed redundant elixir file with resistive Alfven waves.

SimonCan committed:
Reformatted compressible mhd 3d equations.

SimonCan committed:
Minor reformatting.

SimonCan committed:
Added turbulent dynamo elixir.

SimonCan committed:
Added dynamos MHD test.

SimonCan committed:
Typo in docs.

SimonCan committed:
Increased relative tolerance for the dynamo test, as we have random numbers.

SimonCan committed:
Remove so far unused functionality form the resistive MHD equations module.

SimonCan committed:
Changed the format of the docs.

SimonCan committed:
Increased relative tolerance for the MHD dynamo test.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Updated dynamo test case.

SimonCan committed:
Updated dynamo elixir example to make it deterministic by using deterministic seeds
for our random numbers.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Added magnetic energy for resistive MHD equations.

SimonCan committed:
Reformatted mhd equation file.

ranocha commented on src/equations/compressible_mhd_3d.jl:
```diff
@@ -0,0 +1,197 @@
+
+    @unpack eta = equations
+
+    # TODO: use primitive gradients
```
?

SimonCan replied:
Removed.

ranocha reviewed (changes requested):
Thanks! It would be great to get a review from our MHD master @andrewwinters5000

SimonCan committed:
Change description of the physical setup.

SimonCan committed:
Removed relative errors and reformatted.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl:
```diff
@@ -0,0 +1,96 @@
+
+time_int_tol = 1e-5
+sol = solve(ode, RDPK3SpFSAL49(), abstol=time_int_tol, reltol=time_int_tol, dt = 1e-5,
+            save_everystep=false, callback=callbacks)
```
Decided for the `StepsizeCallback`.

SimonCan committed:
Added link to Pencil Code example.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+# A turbulent helically driven dynamo model in resistive compressible MHD.
+# We start with a weak small-scale magnetic field provided as Gaussian noise.
+# The energy is supplied via a forcing function f_force(x, t), which is delta-correlated in time.
+# For a comparison look into the PC run pencil-code/simon/forced/alpha2_periodic/64_kf_5_a.
```
That run is not public. However, I have added a public sample.

SimonCan committed:
Changed UInt64 to target type.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+function initial_condition_gaussian_noise_mhd(x, t, equations)
+    amplitude = 1e-8
+
+    seed = reinterpret(UInt64, t + x[1] + x[2] + x[3])
```
Changed it to target unsigned integer type `UInt`. Since this defines the seed of a random number the results are not binary reproducible on different systems.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+    rho_v2 = 0.0
+    rho_v3 = 0.0
+    rho_e = 10.0
+    B1 = randn(1)[1] * amplitude
+    B2 = randn(1)[1] * amplitude
+    B3 = randn(1)[1] * amplitude
```
Do you mean that for the same seeds we could get different rando numbers using different Julia versions? That would be an issue for the test.  Do we allow distinctions in the tests with regard to the Julia version? That way we could just add an `if` condition.

ranocha replied:
Yes, that's what I mean. I would detest a statement like `if VERSION ...` in code like this. You may look at https://github.com/JuliaRandom/StableRNGs.jl for something that is less efficient but stable. But I would still like to know whether there is another way to initialize this setup and get a review from other core devs, e.g., @andrewwinters5000

SimonCan replied:
`StableRNG` looks promising. However, they might not be entirely free from update issues:
_The_ StableRNG type provided by this package strives for stability, but if bugs which require breaking this promise are found, a new major version will be released with the fix._
Considering their library consists of a single file, maybe we can include it as third party software in Trixi.jl. That way we would have control over the version. But that is a serious design question.

ranocha replied:
I wouldn't include it. They basically say that that only change the stream if it's crap due to bugs. In this case, we definitely want to update our tests anyway.

ranocha replied:
And they say it will be a new major version, so it will not hit us unexpected but can be controlled by `test/Project.toml` compatibility settings.

SimonCan replied:
Seems like StableRNGs is a viable solution. Since we specify the version number we should not have any unexpected test breakage. Just pushed my changes.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Hendrik Ranocha <ranocha@users.noreply.github.com>

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Hendrik Ranocha <ranocha@users.noreply.github.com>

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+    k = [0.0, 0.0, 0.0]
+    k[1] = (rand(1)[1] * 2 - 1) * pi
+    k[2] = (rand(1)[1] * 2 - 1) * pi
+    k[3] = (rand(1)[1] * 2 - 1) * pi
```
Corrected to `rand()`.

SimonCan committed:
Changed randn(1)[1] to randn.

SimonCan committed:
Added SVector to avoid memory allocation.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+    ee[2] = rand(1)[1] * 2 - 1
+    ee[3] = rand(1)[1] * 2 - 1
+    # Normalize the vector e.
+    ee = ee / norm(ee)
```
Changed to `ee = SVector(rand(), rand(), rand()) .* 2 .- 1`.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Hendrik Ranocha <ranocha@users.noreply.github.com>

SimonCan committed:
Fixed description.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+
+
+###############################################################################
+# semidiscretization of the ideal compressible MHD equations
```
Done.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Removed rtol and atol.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,167 @@
+
+time_int_tol = 1e-5
+sol = solve(ode, RDPK3SpFSAL49(), abstol = time_int_tol, reltol = time_int_tol, dt = 1e-5,
+            save_everystep = false, callback = callbacks)
```
Fixed.

SimonCan committed:
Changed indendation of doc string.

SimonCan commented on src/equations/compressible_mhd_3d.jl:
```diff
@@ -0,0 +1,197 @@
+@doc raw"""
+CompressibleMhdDiffusion3D(gamma, inv_gamma_minus_one,
+μ, Pr, eta, kappa,
+equations, gradient_variables)
```
Done.

SimonCan committed:
Update src/equations/compressible_mhd_3d.jl

Co-authored-by: Hendrik Ranocha <ranocha@users.noreply.github.com>

SimonCan committed:
Removed commented code.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Removed comment.

SimonCan committed:
Updated dynamo test results.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Added StableRNGs.

ranocha commented on Project.toml:
We don't use StableRNGs.jl in Trixi.jl itself - only in the tests, don't we? Thus, please do not change the main `Project.toml` but the one in the `test` directory.

SimonCan replied:
Done.

ranocha commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,164 @@
+    rng = StableRNG(seed)
+
+    # Random phase -pi < phi <= pi
+    phi = (rand() * 2 - 1) * pi
```
More `rand`

SimonCan replied:
Fixed.

ranocha commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,164 @@
+    phi = (rand() * 2 - 1) * pi
+
+    # Random vector k, also delta-correlated in time.
+    k = [0.0, 0.0, 0.0]
+    k[1] = (randn(rng) * 2 - 1) * pi
+    k[2] = (randn(rng) * 2 - 1) * pi
+    k[3] = (randn(rng) * 2 - 1) * pi
+    k = k / norm(k)
+    k = 6 * k
+    k_norm = k / norm(k)
```
That looks like an unnecessary allocation

SimonCan replied:
Done.

ranocha commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,164 @@
+    k_norm = k / norm(k)
+
+    # Random unit vector, not aligned with k.
+    ee = SVector(rand(), rand(), rand()) .* 2 .- 1
```
`rand`

SimonCan replied:
Done.

SimonCan committed:
Moved StableRNG from the main Project.toml to the test Project.toml.

SimonCan committed:
Changed some rand() functions to randn(rng).
Updated mhd dynamo test accordingly.

andrewwinters5000 commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl:
```diff
@@ -0,0 +1,95 @@
+                initial_refinement_level = 3,
+                n_cells_max = 50_000) # set maximum capacity of tree data structure
+
+function initial_condition_constant_alfven(x, t, equations)
```
Is there a reference that could be added for this (simplified) Alfvén wave test case, e.g., the FLASH manual?

andrewwinters5000 commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl:
```diff
@@ -0,0 +1,95 @@
+# ODE solvers, callbacks etc.
+
+# Create ODE problem with time span `tspan`
+tspan = (0.0, 0.1)
```
Why is the final time so short? Usually the Alfvén test needs to run to integer end times for error computation due to the solution periodicity.

SimonCan replied:
Changed it to 1.0.

andrewwinters5000 commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+
+prandtl_number() = 0.72
+# Corresponds to previous alpha^2 runs.
+mu() = 2e-3
+eta = 2e-3
```
The PENCIL code link given at the beginning of this file recommends these quantities to be `5e-3` why pick different values?

andrewwinters5000 commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+# ODE solvers, callbacks etc.
+
+# Create ODE problem with time span `tspan`
+tspan = (0.0, 0.001)
```
This final time also seems very short especially for a turbulent box. What is true desired final time? If this shortened time is just for the testing you can pass `tspan=(0.0, 0.001)` to the appropriate `@test_trixi_include` call.

SimonCan replied:
For a turbulent dynamo we would need a rather long time. Unrealistically long for tests. In the spirit of other tests, I keep this time short.

andrewwinters5000 replied:
I totally understand that. What is the target final time for the test case? You could add a comment here that says what you wrote above. One `tspan` is artificially short for testing /  debugging and then a commented `tspan` could be labeled as the "real" time span for the turbulent box.

andrewwinters5000 replied:
Why was this resolved? You did not answer my question. What is the intended final time of this test case? Can it be found in the paper you referenced above? If that is the case, point the user to this information. Right now, this elixir seems nonsensical to invest all this effort to only integrate until a final time of 0.001.

As I mentioned in a previous comment, you can set a shortened end time within the `trixi_test` call. There is not necessarily any need to set this shorter time within the actual elixir file.

SimonCan replied:
The time span has been changed to 10.0 in the elixir and to 0.001 in the test.

andrewwinters5000 commented on src/equations/compressible_mhd_3d.jl:
```diff
@@ -0,0 +1,185 @@
+Fluid properties such as the dynamic viscosity $\mu$ and magnetic diffusion $\eta$
+can be provided in any consistent unit system, e.g.,
+[$\mu$] = kg m⁻¹ s⁻¹.
+
```
This docstring is incomplete. It should include the form of the additional visco-resistive fluxes. Also maybe point to one of the papers from Andrés for the particular form and scaling of the governing equations.

SimonCan replied:
Added equations in the docs and a reference to Andre's paper.

andrewwinters5000 commented on src/equations/compressible_mhd_3d.jl:
```suggestion
        f5 = (v1 * tau_11 + v2 * tau_12 + v3 * tau_13 + q1) * mu +
             (B2 * (dB2dx - dB1dy) + B3 * (dB3dx - dB1dz)) * eta
        f6 = zero(rho)
        f7 = eta * (dB2dx - dB1dy)
        f8 = eta * (dB3dx - dB1dz)
```
All of the new resistive terms are missing several components. I updated each Cartesian flux accordingly. This is how one would write them down on paper so they are likely computationally inefficient.

andrewwinters5000 commented on src/equations/compressible_mhd_3d.jl:
```suggestion
        g5 = (v1 * tau_12 + v2 * tau_22 + v3 * tau_23 + q2) * mu +
             (B1 * (dB1dy - dB2dx) + B3 * (dB3dy - dB2dz)) * eta
        g6 = eta * (dB1dy - dB2dx)
        g7 = zero(rho)
        g8 = eta * (dB3dy - dB2dz)
```

andrewwinters5000 commented on src/equations/compressible_mhd_3d.jl:
```suggestion
        h5 = (v1 * tau_13 + v2 * tau_23 + v3 * tau_33 + q3) * mu +
             (B1 * (dB1dz - dB3dx) + B2 * (dB2dz - dB3dy) * eta
        h6 = eta * (dB1dz - dB3dx)
        h7 = eta * (dB2dz - dB3dy)
        h8 = zero(rho)
```

andrewwinters5000 commented on test/test_tree_3d_mhd.jl:
```diff
@@ -81,6 +81,18 @@ EXAMPLES_DIR = pkgdir(Trixi, "examples", "tree_3d_dgsem")
       l2   = [0.0186712969755079, 0.01620736832264799, 0.01620736832264803, 0.016207474382769683, 0.07306422729650594, 0.007355137041002365, 0.0073551370410023425, 0.00735520932001833, 0.000506140942330923],
       linf = [0.28040713666979633, 0.27212885844703694, 0.2721288584470349, 0.2837380205051839, 0.7915852408267114, 0.08770240288089526, 0.08770240288089792, 0.08773409387876674, 0.050221095224119834])
   end
+
```
These test values will need updated once the corrected viscous fluxes are implemented.

SimonCan replied:
They have been updated.

andrewwinters5000 reviewed (changes requested):
Thanks for tackling this @SimonCan ! There is a bit of cleanup that needs done.

andrewwinters5000 commented on src/equations/compressible_mhd_3d.jl:
```diff
@@ -0,0 +1,185 @@
+@doc raw"""
```
Also, I just realized looking at this file again (as well as both Navier-Stokes files) that it is missing the `@muladd` header
```julia
# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
# Since these FMAs can increase the performance of many numerical algorithms,
# we need to opt-in explicitly.
# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
@muladd begin
#! format: noindent

...

end # @muladd
```
This should be added as well.

SimonCan replied:
I changed the name of the files and equations. Still need to check that I did it everywhere I should have.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update src/equations/compressible_mhd_3d.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update src/equations/compressible_mhd_3d.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Added Alfven reference.

SimonCan committed:
Changed end time for MHD wagve elixir.

SimonCan committed:
Changed k-vector into SVector.

SimonCan committed:
Added reference for dynamo forcing.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+end
+
+"""
+    source_terms_helical_forcing(u, x, t, equations::CompressibleMhdDiffusion3D)
```
Just added a paper for reference.

SimonCan committed:
Added DOI for dynamo model.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+# A turbulent helically driven dynamo model in resistive compressible MHD.
+# We start with a weak small-scale magnetic field provided as Gaussian noise.
+# The energy is supplied via a forcing function f_force(x, t), which is delta-correlated in time.
+# For a comparison look into the Pencil Code example https://github.com/pencil-code/pencil-code/tree/master/samples/helical-MHDturb.
```
Since we talk about the entire model (equations, initial condition and forcing) I prefer to keep it here.
But I added a sentence to the initial condition.

andrewwinters5000 replied:
Okay, that makes sense.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update src/equations/compressible_mhd_3d.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Added muladd header and footer.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Renamed visco-resistive MHD equations.

SimonCan committed:
Removed redundant comment.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+# semidiscretization of the resistive compressible MHD equations
+
+prandtl_number() = 0.72
+# Corresponds to previous alpha^2 runs.
```
This comment is redundant.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+mu() = 2e-3
+eta = 2e-3
+
+gamma = 1.0 + 2.0 / 3.0
```
1 + degrees of freedom / dimensions.

andrewwinters5000 replied:
Aha, so this is related to the comment of a monotomic gas in 2d having `gamma = 1.0 + 2.0 / 2.0 = 2.0`. Perhaps here add a comment that this is a monotomic gas in 3d.

SimonCan committed:
Changed et to eta().

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,165 @@
+
+equations = IdealGlmMhdEquations3D(gamma)
+equations_parabolic = CompressibleMhdDiffusion3D(equations, mu = mu(),
+                                                 Prandtl = prandtl_number(), eta = eta,
```
Done.

SimonCan committed:
Changed nu and eta for dynamo test.

SimonCan committed:
Updated dynamo test results.

SimonCan commented:
After the latest change the example elixir `examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl`
gives me this error message during run time after a few time steps:
```
ERROR: LoadError: DomainError with -2.8765548088136064:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
DomainError detected in the user `f` function. This occurs when the domain of a function is violated.
For example, `log(-1.0)` is undefined because `log` of a real number is defined to only output real
numbers, but `log` of a negative number is complex valued and therefore Julia throws a DomainError
by default. Cases to be aware of include:

* `log(x)`, `sqrt(x)`, `cbrt(x)`, etc. where `x<0`
* `x^y` for `x<0` floating point `y` (example: `(-1.0)^(1/2) == im`
```

Is this a typical error with a typical source? If not I will need to do some thorough debugging.

SimonCan committed:
Minor typos.

andrewwinters5000 commented:
> Is this a typical error with a typical source? If not I will need to do some thorough debugging.

My best guess is that your explicit time step is too large. Since the viscous fluxes have changed you may have to adjust the `cfl` value. Also, the `StepsizeCallback` only estimates the eigenvalues, and in turn the CFL condition, for the advective components. The viscous time step tends to be stiffer and an estimate of the viscous time step restriction needs to be implemented for an alternative `StepsizeCallback` that could dispatch on both `equations` and `equations_parabolic`. @ranocha, @sloede: Should we add such a time step estimate feature to the parabolic TODO list #1147?

SimonCan committed:
Adjusted cfl condition for MHD elixir.

SimonCan committed:
Corrected test values for damped Alfven wave after adjusting the CFL condition.

SimonCan commented:
After the latest change the example elixir `examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl`
gives me this error message during run time after a few time steps:
```
ERROR: LoadError: DomainError with -2.8765548088136064:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
DomainError detected in the user `f` function. This occurs when the domain of a function is violated.
For example, `log(-1.0)` is undefined because `log` of a real number is defined to only output real
numbers, but `log` of a negative number is complex valued and therefore Julia throws a DomainError
by default. Cases to be aware of include:

* `log(x)`, `sqrt(x)`, `cbrt(x)`, etc. where `x<0`
* `x^y` for `x<0` floating point `y` (example: `(-1.0)^(1/2) == im`
```

Is this a typical error with a typical source? If not I will need to do some thorough debugging.

> > Is this a typical error with a typical source? If not I will need to do some thorough debugging.
> 
> My best guess is that your explicit time step is too large. Since the viscous fluxes have changed you may have to adjust the `cfl` value. Also, the `StepsizeCallback` only estimates the eigenvalues, and in turn the CFL condition, for the advective components. The viscous time step tends to be stiffer and an estimate of the viscous time step restriction needs to be implemented for an alternative `StepsizeCallback` that could dispatch on both `equations` and `equations_parabolic`. @ranocha, @sloede: Should we add such a time step estimate feature to the parabolic TODO list #1147?

Excellent. That fixed the problem. I changed cfl = 1.5 to half of that: cfl = 0.75.

SimonCan committed:
Formatting.

SimonCan reviewed (commented):
@andrewwinters5000 : Not sure where these changes can be derived. The flux form of the resistive MHD part should have been correct.

SimonCan reviewed (commented):
The diffusive part of the induction equation should just have this simple form:
(∂t B)_diff = ∂x(∂x B) + ∂y (∂y B) + ∂z (∂z B)

andrewwinters5000 commented:
> @andrewwinters5000 : Not sure where these changes can be derived. The flux form of the resistive MHD part should have been correct.

I looked at the governing equations from one of [Andrés' papers](https://arxiv.org/pdf/2012.12040.pdf) (see Section 2.2.1). Here is a screenshot of the viscous fluxes where `eta` in the current implementation plays the role of \mu_R /  \mu_0 in the notation of Andrés.

<img width="343" alt="Screenshot 2023-08-03 at 12 43 15" src="https://github.com/trixi-framework/Trixi.jl/assets/25242486/7c90bf4d-c281-4d4b-beb0-ec812fc26a69">

SimonCan commented:
> > The diffusive part of the induction equation should just have this simple form: (∂t B)_diff = ∂x(∂x B) + ∂y (∂y B) + ∂z (∂z B)
> 
> I don't think it is just a simple Laplacian. From what I remember the diffusion part in the induction equation involves curl( curl(B) ) which we can rewrite as grad(div(B)) - div(grad(B)).

I see now. We assume div(B) is not 0 and need to factor in those terms. Then we get the additional term grad(div(B)) which has the opposite sign of Laplace(B).

SimonCan committed:
Added the visco-resistive equations in the documentation.

andrewwinters5000 commented on src/equations/equations_parabolic.jl:
```diff
@@ -13,3 +13,8 @@ abstract type AbstractCompressibleNavierStokesDiffusion{NDIMS, NVARS} <:
+# Resistive and visuous MHD
+abstract type AbstractViscoResistiveMhd{NDIMS, NVARS} <:
+              AbstractEquationsParabolic{NDIMS, NVARS} end
+include("visco_resistive_mhd_3d.jl")
```
I would add to some kind of TODO list or a student project ideas list that the 2D implementation of the visco-resistive equations could be interesting. What do you think @sloede ?

SimonCan replied:
I increased the simulation time to 10, which is rounded from the 8.33 of the paper. Since such a run would take several hours I changed the time span in the test,

andrewwinters5000 commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+where `\tau` is the viscous stress tensor and `q = \kappa \overrightarrow{\nabla} T`.
+For the induction term we have the usual Laplace operator on the magnetic field
+but we also include terms with `div(B)`.
+Divergence cleaning is done using the `\Psi` field.
```
```suggestion
Divergence cleaning is done using the `\psi` field.
```

andrewwinters5000 commented on src/equations/visco_resistive_mhd_3d.jl:
````diff
@@ -0,0 +1,210 @@
+can be provided in any consistent unit system, e.g.,
+[$\mu$] = kg m⁻¹ s⁻¹.
+
+The equations given here are the visco-resistive part of the MHD equations
+in conservation form:
+```math
+\overleftrightarrow{f}^\mathrm{\mu\eta} =
+\begin{pmatrix}
+\overrightarrow{0} \\
+\underline{\tau} \\
+\underline{\tau}\overrightarrow{v} - \overrightarrow{\nabla}q -
+ \eta\left( (\overrightarrow{\nabla}\times\overrightarrow{B})\times\overrightarrow{B} \right) \\
+\eta \left( (\overrightarrow{\nabla}\overrightarrow{B})^\mathrm{T} - \overrightarrow{\nabla}\overrightarrow{B} \right) \\
+\overrightarrow{0}
+\end{pmatrix},
+```
````
This looks fine, but I realized after my previous comment that the ideal GLM-MHD equations do not contain proper docstrings for the equation formulation and their description. Therefore, I think we can save a "proper" docstring writing for the ideal GLM-MHD equations as well as the new visco-resistive GLM-MHD equations for a different PR. This way we can be sure that the notation across all files is consistent and easy to review. What do you think @sloede @ranocha ?

SimonCan replied:
Sounds good. Maybe we can keep this one and change it in case we need to.

ranocha replied:
:+1:

andrewwinters5000 reviewed (dismissed):
Just a few more comments about the documentation.

SimonCan committed:
Added a few comments on Alfven wave setup.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl:
```diff
@@ -0,0 +1,96 @@
+                n_cells_max = 50_000) # set maximum capacity of tree data structure
+
+function initial_condition_constant_alfven(x, t, equations)
+    # Alfven H., 150, p. 450, Nature (1942), DOI: 10.1038/150405d0
```
The setup does not come from any particular source. It's just a simple perturbation. I added two more references that should clarify things.

SimonCan committed:
Added section reference for paper.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,163 @@
+# We start with a weak small-scale magnetic field provided as Gaussian noise.
+# The energy is supplied via a forcing function f_force(x, t), which is delta-correlated in time.
+# For a comparison look into the Pencil Code example https://github.com/pencil-code/pencil-code/tree/master/samples/helical-MHDturb.
+# For more information see Cnadelaresi et al. (2013), DOI: 10.1103/PhysRevE.87.043104.
```
Added the relevant section.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Reformatted reference.

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,163 @@
+"""
+    initial_condition_gaussian_noise_mhd(x, t, equations)
+
+Define the initial condition with a weak small-scale Gaussian noise magnetic field.
```
Might be overkill, as the paper does not formalize the condition.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_dynamo.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Update src/equations/visco_resistive_mhd_3d.jl

Co-authored-by: Andrew Winters <andrew.ross.winters@liu.se>

SimonCan committed:
Increased the final time for the MHD dynamo to 10.0.

SimonCan committed:
Changed time span for MHD dynamo test.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

ranocha commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -2,7 +2,7 @@
 # The energy is supplied via a forcing function f_force(x, t), which is delta-correlated in time.
 # For a comparison look into the Pencil Code example https://github.com/pencil-code/pencil-code/tree/master/samples/helical-MHDturb.
-# For more information see Cnadelaresi et al. (2013), DOI: 10.1103/PhysRevE.87.043104.
+# For more information see Cnadelaresi et al. (2013), DOI: 10.1103/PhysRevE.87.043104 section II A.
```
```suggestion
# For more information see Candelaresi et al. (2013), DOI: 10.1103/PhysRevE.87.043104 section II A.
```
?

SimonCan replied:
This already got corrected, thanks.

ranocha commented:
> The viscous time step tends to be stiffer and an estimate of the viscous time step restriction needs to be implemented for an alternative `StepsizeCallback` that could dispatch on both `equations` and `equations_parabolic`. @ranocha, @sloede: Should we add such a time step estimate feature to the parabolic TODO list #1147?

Done :+1:

SimonCan committed:
Added convergence test to the diffusive Alfven wave elixir example.

SimonCan committed:
Changed the diffusive Alfven wave parameters.

SimonCan committed:
Updated test results for diffusive Alfven waves.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Attempt to get convergence for MHD.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
New MHD convergence tests.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Updated viscous MHD test results.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Removed Andrew's constructed solution and reinstated simpler Alfven waves.
Added optional viscosity and resitivity parameters.

SimonCan committed:
Fixed convergen issue with visco-resistive MHD.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Tried further 3d convergence tests using more complex manufactured solutions.

SimonCan committed:
Added 3d Alfven wave for convergence test.

SimonCan committed:
Corrected 3d Alfven waves for convergence test.

SimonCan committed:
Removed 1d Alfven test.
x

SimonCan committed:
Autoformatter on visco-resistive Walfven waves elixir.

SimonCan committed:
Added convergence test with non-trivial density truncation error.

SimonCan committed:
Corrected local truncation error for visco-resistive MHD convergence test.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

DanielDoehring commented on src/Trixi.jl:
```diff
@@ -140,6 +140,7 @@ export AcousticPerturbationEquations2D,
        CompressibleEulerMulticomponentEquations1D,
        CompressibleEulerMulticomponentEquations2D,
        IdealGlmMhdEquations1D, IdealGlmMhdEquations2D, IdealGlmMhdEquations3D,
+       ViscoResistiveMhd3D,
```
```suggestion
       ViscoResistiveMhdEquations3D,
```

(for consistency with the other names)

DanielDoehring commented on src/equations/equations_parabolic.jl:
```diff
@@ -21,3 +21,8 @@ include("compressible_navier_stokes.jl")
 include("compressible_navier_stokes_3d.jl")
+
+# Resistive and visuous MHD
+abstract type AbstractViscoResistiveMhd{NDIMS, NVARS} <:
```
```suggestion
abstract type AbstractViscoResistiveMhdEquations{NDIMS, NVARS} <:
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+#! format: noindent
+
+@doc raw"""
+    ViscoResistiveMhd3D(gamma, inv_gamma_minus_one,
```
```suggestion
    ViscoResistiveMhdEquations3D(gamma, inv_gamma_minus_one,
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+
+For more details see e.g. arXiv:2012.12040.
+"""
+struct ViscoResistiveMhd3D{GradientVariables, RealT <: Real,
```
```suggestion
struct ViscoResistiveMhdEquations3D{GradientVariables, RealT <: Real,
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+"""
+struct ViscoResistiveMhd3D{GradientVariables, RealT <: Real,
+                           E <: AbstractIdealGlmMhdEquations{3}} <:
+       AbstractViscoResistiveMhd{3, 9}
```
```suggestion
       AbstractViscoResistiveMhdEquations{3, 9}
```

RZywczok replied:
ViscoResistiveMhdDiffusion3D{3, 9, GradientVariables}

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+end
+
+# default to primitive gradient variables
+function ViscoResistiveMhd3D(equations::IdealGlmMhdEquations3D;
```
```suggestion
function ViscoResistiveMhdEquations3D(equations::IdealGlmMhdEquations3D;
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+    # Important note! Factor of μ is accounted for later in `flux`.
+    kappa = gamma * inv_gamma_minus_one / Pr
+
+    ViscoResistiveMhd3D{typeof(gradient_variables), typeof(gamma), typeof(equations)
```
```suggestion
    ViscoResistiveMhdEquations3D{typeof(gradient_variables), typeof(gamma), typeof(equations)
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+# of the paper by Rueda-Ramírez, Hennemann, Hindenlang, Winters, and Gassner
+# "An Entropy Stable Nodal Discontinuous Galerkin Method for the resistive
+#  MHD Equations. Part II: Subcell Finite Volume Shock Capturing"
+function flux(u, gradients, orientation::Integer, equations::ViscoResistiveMhd3D)
```
```suggestion
function flux(u, gradients, orientation::Integer, equations::ViscoResistiveMhdEquations3D)
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```suggestion
                                                  equations::ViscoResistiveMhdEquations3D{
```

DanielDoehring commented on src/equations/visco_resistive_mhd_3d.jl:
```suggestion
                                                 ::ViscoResistiveMhdEquations3D{
```

DanielDoehring commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,171 @@
+gamma = 1.0 + 2.0 / 3.0
+
+equations = IdealGlmMhdEquations3D(gamma)
+equations_parabolic = ViscoResistiveMhd3D(equations, mu = mu(),
```
```suggestion
equations_parabolic = ViscoResistiveMhdEquations3D(equations, mu = mu(),
```

(see below)

DanielDoehring commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,171 @@
+  Kinetic helicity needed to drive large-scale dynamos
+  [DOI: 10.1103/PhysRevE.87.043104](https://doi.org/10.1103/PhysRevE.87.043104)
+"""
+function source_terms_helical_forcing(u, x, t, equations::IdealGlmMhdEquations3D)
```
```suggestion
function source_terms_helical_forcing(u, x, t, equations::ViscoResistiveMhdEquations3D)
```

DanielDoehring commented on examples/tree_3d_dgsem/elixir_mhd_dynamo.jl:
```diff
@@ -0,0 +1,171 @@
+end
+
+"""
+    source_terms_helical_forcing(u, x, t, equations::ViscoResistiveMhd3D)
```
```suggestion
    source_terms_helical_forcing(u, x, t, equations::ViscoResistiveMhdEquations3D)
```

DanielDoehring commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl:
```diff
@@ -0,0 +1,124 @@
+eta_const = 2e-2
+
+equations = IdealGlmMhdEquations3D(5 / 3)
+equations_parabolic = ViscoResistiveMhd3D(equations, mu = mu_const,
```
```suggestion
equations_parabolic = ViscoResistiveMhdEquations3D(equations, mu = mu_const,
```

DanielDoehring reviewed (changes requested):
For consistency with the other parabolic terms, see 

https://github.com/trixi-framework/Trixi.jl/blob/3c1762b20189cba6b27abaa88e1575dc495c5c73/src/equations/equations_parabolic.jl#L18-L19

and 

https://github.com/trixi-framework/Trixi.jl/blob/3c1762b20189cba6b27abaa88e1575dc495c5c73/src/equations/compressible_navier_stokes_1d.jl#L83

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan commented:
> For consistency with the other parabolic terms, see
> 
> https://github.com/trixi-framework/Trixi.jl/blob/3c1762b20189cba6b27abaa88e1575dc495c5c73/src/equations/equations_parabolic.jl#L18-L19
> 
> and
> 
> https://github.com/trixi-framework/Trixi.jl/blob/3c1762b20189cba6b27abaa88e1575dc495c5c73/src/equations/compressible_navier_stokes_1d.jl#L83

Not sure about this, as visco-resistive already implies diffusion. Even f it would bring it in line with the names of the Navier-Stokes equations.

SimonCan committed:
Change parametric type of supertype of visco-resistive MHD equations.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

DanielDoehring commented on test/test_tree_3d_mhd.jl:
```diff
@@ -291,6 +291,20 @@ end
+    @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_mhd_diffusive_alfven_wave.jl"),
+      l2 = [3.470701487566952e-5, 0.00037320266655784104, 0.0001814791326953306, 0.0004187613798734842, 0.00011587163989250743, 0.00037032036399761947, 0.00018270779781144185, 0.00041229458967006154, 2.7753210639128732e-5],
+      linf = [0.0001348833811495309, 0.000892691474205648, 0.00043846917904940587, 0.0010014206836504308, 0.000443288547495424, 0.0008149498186467707, 0.00042111533550182045, 0.000918156282098577, 0.00010109055017304109])
+  end
```
Here we definitely should check for allocations as for the other tests.

SimonCan replied:
The memory issue has been fixed.

DanielDoehring commented on test/test_tree_3d_mhd.jl:
```diff
@@ -291,6 +291,20 @@ end
+      l2   = [5.136817044406367e-9, 1.0728023104350635e-6, 1.0741472192876857e-6, 1.1384862784905957e-6, 8.505678187506508e-8, 1.2404355188990195e-8, 1.2260968462092833e-8, 1.2453935882908254e-8, 5.406090684284657e-9],
+      linf = [6.223255566428776e-8, 3.476280403710667e-6, 3.2024943277794614e-6, 3.2985189119007034e-6, 1.0206843477078564e-6, 7.224815773523103e-8, 9.16975614740533e-8, 7.92452544029745e-8, 1.201338384265285e-7],
+     tspan = (0.0, 0.001))
+  end
```
Here we definitely should check for allocations as for the other tests.

SimonCan replied:
I removed the MHD dynamo test. Dynamos require a long time to be excited and become fully turbulent. Since we already test the visco-resistive equations in [test/test_tree_3d_mhd.jl](https://github.com/trixi-framework/Trixi.jl/pull/1323/files/d3c90f6208dea269d871650feb0131b503646188#diff-6abf3de2eaa2fb7d2d53a57d2fa307cdf728cb9949862d478f1597dae5ee14d1), we don't need this.

DanielDoehring replied:
Hm not so sure if the others will accept this, usually every elixir has a test.

DanielDoehring replied:
But maybe reduce the `tspan` such that the test does not run longer than 10 sec.

SimonCan replied:
The dynamo example elixir is rather specific. So, I decided to remove it. The Alfven wave example should be sufficient to show case the usage of the visco-resistive equations.

SimonCan committed:
Removed MHD dynamo test due to long run time.
Added memory allocation chck for the visco-resistive MHD test (Alfven waves).

SimonCan committed:
UIpdated errors for visco-resistive MHD test after removing soure term.

SimonCan committed:
Removed source term.

SimonCan committed:
Added visco-resistive MHD converge test.
This is currently work in progress.

jlchan commented:
> > > For consistency with the other parabolic terms, see
> 
> > > https://github.com/trixi-framework/Trixi.jl/blob/3c1762b20189cba6b27abaa88e1575dc495c5c73/src/equations/equations_parabolic.jl#L18-L19
> 
> > > 
> 
> > > and
> 
> > > https://github.com/trixi-framework/Trixi.jl/blob/3c1762b20189cba6b27abaa88e1575dc495c5c73/src/equations/compressible_navier_stokes_1d.jl#L83
> 
> > 
> 
> > Not sure about this, as visco-resistive already implies diffusion. Even f it would bring it in line with the names of the Navier-Stokes equations.
> 
> 
> 
> Opinions on this @andrewwinters5000 @jlchan ?

If it's not too much trouble, I think adding "Diffusion" to make it consistent would be nice. Non experts may not be aware that resistive implies diffusion.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Added 2d Alfven wave test.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

andrewwinters5000 commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_convergence.jl:
```diff
@@ -0,0 +1,148 @@
+    Va = omega / (ny * sqr)
+    phi_alv = omega / ny * (nx * (x[1] - 0.5 * r) + ny * (x[2] - 0.5 * r)) - Va * t
+
+    rho = 1.0 + e*cos(phi_alv + 1.0)
+    v1 = -e * ny * cos(phi_alv) / rho
+    v2 = e * nx * cos(phi_alv) / rho
+    v3 = e * sin(phi_alv) / rho
+    p = 1.0
+    B1 = nx - rho * v1 * sqr
+    B2 = ny - rho * v2 * sqr
+    B3 = -rho * v3 * sqr
+    psi = 0.0
```
This looks like a fairly complicated manufactured solution that would be the ideal MHD Alfvén wave (so no source term required if viscosity and resistivity are set to 0). If I may propose an alternative, here is an MMS from my old (2D) Fortran code:
```fortran
            alpha = 2.0_RP*PI*(x+y)-4.0_RP*t
            Resu(1) = SIN(alpha)+4.0_RP
            Resu(2) = SIN(alpha)+4.0_RP
            Resu(3) = SIN(alpha)+4.0_RP
            Resu(4) = 0.0_RP
            Resu(5) = 2.0_RP*(SIN(alpha)+4.0_RP)**2
            Resu(6) = SIN(alpha)+4.0_RP
            Resu(7) = -SIN(alpha)-4.0_RP
            Resu(8) = 0.0_RP
            Resu(9) = 0.0_RP
```

The associated source term is then (fairly) simple and takes the form:
```fortran
            alpha = 2.0_RP*PI*(x+y)-4.0_RP*t
            phi = SIN(alpha)+4.0_RP
            phi_t = -4.0_RP*COS(alpha)
            phi_x = 2.0_RP*PI*COS(alpha)
            phi_xx = -4.0_RP*PI**2*SIN(alpha)
            source(1) = phi_t+2.0_RP*phi_x
            source(2) = phi_t+4.0_RP*phi*phi_x+phi_x
            source(3) = source(2)
            source(4) = 0.0_RP
            source(5) = 4.0_RP*phi*phi_t+16.0_RP*phi*phi_x-2.0_RP*phi_x-4.0_RP*eta*(phi_x**2+phi*phi_xx)-4.0_RP*mu/Prandtl*phi_xx
            source(6) = source(1)-2.0_RP*eta*phi_xx
            source(7) = -source(6)
            source(8) = 0.0_RP
            source(9) = 0.0_RP
```
Maybe a somewhat simpler MMS for the 3D case could help in this instance.

DanielDoehring replied:
What are domain and final time for this case? Domain probably $[0, 1]^2$?

andrewwinters5000 replied:
Yes, sorry, the domain in [0,1]^2 and the final time I typically used 1.5. The `Resu` is already the conservative variables such that we are setting the velocities to be `(v1_, v_2, v_3) = (1, 1, 0)`.

DanielDoehring replied:
Thanks!

SimonCan replied:
Thanks for that. I had a go with this and tried some modifications, but the convergence rate is close to 0, even without viscosity or magnetic resistivity. I am running to time 0.1, though, as it takes rather long.

andrewwinters5000 commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_convergence.jl:
```diff
@@ -0,0 +1,148 @@
+end
+
+
+@inline function source_terms_mhd_convergence_test(u, x, t, equations)
```
From the looks of it there source term contribution were determined with computer algebra software (maybe Maple?) and then copied into here. It is hard to parse what exactly is going on. From the initial condition routine it seems that you are using the periodic manufactured solution (MMS) from Altmann's thesis. The energy source term is especially ugly. Would it be possible that you pick a simpler MMS ansatz and then compute the source term again (even by hand if necessary)? This is what I did for the compressible Navier-Stokes MMS test.

SimonCan replied:
Yes, that result is from using sympy. I apply the differential operator from the PDE (visco-resistive MHD) on the manufactured solution. There are two test functions. One is a simple Alfven wave in the x-direction and the other is indeed from Altmann's thesis, which is also an Alfven wave. The simpler ansatz is the Alfven wave in x, which fails in the energy term when mu and eta > 0.

SimonCan committed:
Removed dynamo example elixir, as it is too specific/elaborate.

SimonCan committed:
Corrected typo in diffusive mhd Alfven wave test results.

SimonCan committed:
Added Andrew's suggestion for a manufactured solution in diffusive MHD.

SimonCan committed:
Added 2d visco-resistive MHD conergence test.

SimonCan committed:
Added visco-resistive MHD equations in 2d.
Currently without resistivity.
This will be moved to a different PR.

SimonCan committed:
Added visco-resistive MHD in 2d.

SimonCan committed:
Corrected 2d visco-resistive MHD equations (non-conservative part).

SimonCan committed:
Corrected manufactured solution for 2d visco-resistive MHED convergence test.

SimonCan committed:
Added ViscoResistiveMhd2D to the export.

SimonCan committed:
Clean up convergence test elixir for 2d visco-resistive MHD equations.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Auto reformatted MHD example elixirs.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Removed 2d MHD visco-resistive elixir.

SimonCan committed:
Remoed 2d MHD visco-resistive euations.

SimonCan committed:
Removed ViscoResistiveMhd2D.

SimonCan committed:
Removed 2d visco-0resistive MHD equations, as this is now part of a seperate PR/branch.

SimonCan committed:
Cleaned up visco-resistive MHD elixir.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Added finite viscosity and resistivity.

SimonCan committed:
Corrected residuals for visco-resistive MHD in 3d.

SimonCan committed:
Corrected residual for 3d visco-resistive MHD convergence test.

SimonCan committed:
Corrected residual of 3d Alfven waves.

SimonCan committed:
Corrected residual for 3d Alfven wave for the visco-resistive MHD convergence
test.

SimonCan committed:
Apple autoformatter on the lengthy expressions for the residuals.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Simplifie terms for visco-resistive MHD convergence.

SimonCan committed:
Removed small numbers that should be 0. Does not affect convergence numbers.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Further convergence tests.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Add explicit (boundary_condition_periodic, boundary_condition_periodic) pass.

Add periodicity = True, as it is not default anymore.

Make SemidiscretizationHyperbolicParabolic mutable.

SimonCan committed:
git pushMerge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Fixed parameters for speedy convergence tests.

SimonCan committed:
Remove redundant includes.

SimonCan committed:
Add further tests for visco-resistive MHD on tree meshes in 3d.

SimonCan committed:
Add AbstractViscoResistiveMhd.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Use invokelatest when calculating u_exact to keep the test quiet.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Corrected subtyping by adding GradientVariables to AbstractViscoResistiveMhd.

SimonCan committed:
Correct root directory for elixir_mhd_diffusive_convergence.jl test.

SimonCan committed:
Change OrdinaryDiffEq to OrdinaryDiffEqLowStorageRK in elixir.

SimonCan committed:
Update examples/tree_3d_dgsem/elixir_mhd_diffusive_convergence.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SimonCan committed:
Update src/equations/equations_parabolic.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SimonCan committed:
Update src/equations/visco_resistive_mhd_3d.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SimonCan committed:
Update src/equations/visco_resistive_mhd_3d.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SimonCan committed:
Update src/equations/visco_resistive_mhd_3d.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Fix directory to test elixir elixir_mhd_diffusive_convergence.jl.

SimonCan committed:
Correct linf and l2 errors for elixir_mhd_diffusive_alfven_wave.jl test.

SimonCan committed:
Include now OrdinaryDiffEqLowStorageRK.

SimonCan committed:
Correct test return values for elixir_mhd_diffusive_convergence.jl.

Fixed parameters for elixir_mhd_diffusive_convergence.jl.

SimonCan committed:
Fix excessive memory allocation for test elixir_mhd_diffusive_convergence.jl.

SimonCan committed:
Add con2prim functions for ViscoResistiveMhd3D.

Uncomment commented out tests.

SimonCan committed:
Merge branch 'main' into sc/mhd-treemesh-3d

SimonCan committed:
Correct errors for visco-resistive MHD tests in 3d.

SimonCan committed:
Merge branch 'sc/mhd-treemesh-3d' of github.com:trixi-framework/Trixi.jl into sc/mhd-treemesh-3d

SimonCan committed:
Removed redundant routines.

SimonCan committed:
Apply autoformatter.

SimonCan committed:
Update src/semidiscretization/semidiscretization_hyperbolic_parabolic.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

SimonCan commented on examples/tree_3d_dgsem/elixir_mhd_diffusive_alfven_wave.jl:
```diff
@@ -0,0 +1,124 @@
+eta_const = 2e-2
+
+equations = IdealGlmMhdEquations3D(5 / 3)
+equations_parabolic = ViscoResistiveMhd3D(equations, mu = mu_const,
```
"Diffusion" would be redundant here, as resisistivity already implies it.

SimonCan commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+#! format: noindent
+
+@doc raw"""
+    ViscoResistiveMhd3D(gamma, inv_gamma_minus_one,
```
Same as above.

SimonCan commented on src/equations/visco_resistive_mhd_3d.jl:
```diff
@@ -0,0 +1,210 @@
+
+For more details see e.g. arXiv:2012.12040.
+"""
+struct ViscoResistiveMhd3D{GradientVariables, RealT <: Real,
```
Same as above.

SimonCan reviewed (commented):
Replied to comments regarding the naming of the implemented equations.
